### DISCUSSION TITLE
Logging incoming requests and returned responses

### DISCUSSION THREAD

**Discussion by tr-gis** [fastapi/fastapi#6345] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.middleware("http")
async def log_stuff(request: Request, call_next):
    logger.debug(f"{request.method} {request.url}")
    response = await call_next(request)
    logger.debug(response)
    logger.debug(response.status_code)
    return response
```

### Description

I want to log all the requests received by FastAPI and the responses that were returned to them.  Initially I am just trying to log them to console , later I will switch it to logging in a file. The end goal is to have a log file that I can use `tail` command to monitor and see if things are going smoothly. I took the middleware approach and put in statements to log the incoming requests and response status code . It works for far, next I want to see the actual json being returned. I am not sure how to get this done. I tried `logger.debug(response.content)` and got` AttributeError: 'StreamingResponse' object has no attribute 'content'`

Sorry I had to redact some stuff in the below image. 
![s](https://user-images.githubusercontent.com/12557720/158164312-799b2547-dc47-4e81-8be2-979550f54f4d.png)

I feel like there should be a better way to get this logging done since most users will have similar requirements. Kindly let me know if there is a better way to go about it.

### Operating System

Linux

### Operating System Details

Linux XXX-desktop 5.13.0-35-generic #40~20.04.1-Ubuntu SMP Mon Mar 7 09:18:32 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### FastAPI Version

0.70.0

### Python Version

Python 3.8.10

**Comment by tr-gis** [comment]
I did could not figure out middleware to get logging done, however I modified the example code under 

> 'Custom APIRoute class in a router' 

 [ section here ](https://fastapi.tiangolo.com/advanced/custom-request-and-route/#accessing-the-request-body-in-an-exception-handler) to get my logging requirements done. Looks like this:

```
class log_stuff(APIRoute):
    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request) -> Response:
            req = await request.body()
            log.info(req)
            response: Response = await original_route_handler(request)
            log.info(response.body)
            return response

        return custom_route_handler
```

**Comment by Roymprog** [comment, marked as answer]
For those looking to implement this, there's a good [SO post](https://stackoverflow.com/a/73464007/10100205) describing two implementations. It's probably best to go with the API route one, as the middleware route has some constraints: ["The unique condition is that this should be placed as the last middleware (the first to be loaded into the app, take care about the reverse ordering), if not it will freeze."](https://github.com/tiangolo/fastapi/issues/394#issuecomment-932721992), plus it's [discouraged by the author of Starlette](https://github.com/encode/starlette/issues/495#issuecomment-494008175).

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue 👍
